container-search/src/main/java/com/yahoo/search/cluster/ClusterMonitor.java

```diff
@@ -74,30 +105,48 @@ public void add(T node, boolean internal) {
 
     /** Called from ClusterSearcher/NodeManager when a node failed */
     public synchronized void failed(T node, ErrorMessage error) {
-        if (closed.get()) return; // Do not touch state if close has started.
-        TrafficNodeMonitor<T> monitor = nodeMonitors.get(node);
-        Boolean wasWorking = monitor.isKnownWorking();
-        monitor.failed(error);
-        if (wasWorking != monitor.isKnownWorking())
-            nodeManager.failed(node);
+        updateMonitoredNode(node, monitor -> monitor.failed(error), nodeManager::failed);
     }
 
     /** Called when a node responded */
     public synchronized void responded(T node) {
-        if (closed.get()) return; // Do not touch state if close has started.
+        updateMonitoredNode(node, TrafficNodeMonitor::responded, nodeManager::working);
+    }
+
+    private void updateMonitoredNode(T node, Consumer<TrafficNodeMonitor<T>> monitorUpdate, Consumer<T> nodeUpdate) {
         TrafficNodeMonitor<T> monitor = nodeMonitors.get(node);
-        Boolean wasWorking = monitor.isKnownWorking();
-        monitor.responded();
-        if (wasWorking != monitor.isKnownWorking())
-            nodeManager.working(node);
+
+        // Don't touch state during shutdown.
+        if (closed.get()) monitor = null;
+
+        // Node was removed during reconfiguration, and should no longer be monitored.
+        if (nodesToRemove.remove(node)) {
+            nodeMonitors.remove(node);
+            monitor = null;
+        }
+
+        // Update monitor state only when it actually changes.
+        if (monitor != null) {
+            Boolean wasWorking = monitor.isKnownWorking();
+            monitorUpdate.accept(monitor);
+            if (wasWorking != monitor.isKnownWorking())
```

Comment:
`monitorUpdate.accept()` changes result of `isKnownWorking()`?

Reply:
Yes, I kept this code as it was. The cluster monitor is used elsewhere, and I suppose we want to avoid unnecessary updates to state there.